### PR TITLE
Reverse sorting on /users/nodes/ by title gives 500 [OSF-5237]

### PR DESCRIPTION
**Pull request by KAsante95**
# Purpose

Enable reverse sorting based on user node attributes such as title and date_created

Issue:
https://openscience.atlassian.net/browse/OSF-5237

# Code

Removed fields.pop(0) and simply selected the first index via fields[0].
I then included a reverse sort for the case in which the first character in the field was a '-', thus indicating a reverse sort. To implement the reverse sort, I simply switched the places of the 1 and -1 from the original sort.

### CODE REVIEW

File: api_tests/base/test_filters.py [hunk 1]
```diff
@@ -222,3 +229,51 @@ def test_convert_value_float(self):
...
+    def test_filter_queryset_fails(self):
+        class query:
+            title = ' '
+            def __init__(self, title):
+                self.title = title
+            def __str__(self):
+                return self.title
+        query_s = []
+        q1 = query('NewProj')
+        query_s.append((q1))
```
Comment (whit537) [inline api_tests/base/test_filters.py]: What's up with the extra parens here? Let's drop them here and throughout these two new tests.

File: api_tests/base/test_filters.py [hunk 2]
```diff
@@ -222,3 +229,51 @@ def test_convert_value_float(self):
...
+        q2 = query('Zip')
+        query_s.append((q2))
+        q3 = query('Proj')
+        query_s.append((q3))
+        q4 = query('Activity')
+        query_s.append((q4))
+        sort_q = sorted(query_s, cmp=filters.sort_multiple(['-title']))
+        sortt = []
+        for i in sort_q:
+            sortt.append(str(i))
```
Comment (whit537) [inline api_tests/base/test_filters.py]: This is a great place for a list comprehension:

```python
sortt = [str(i) for i in sort_q]
```

**Comment by whit537** [comment]
#### Punchlist

 - [ ] remove extra parens—https://github.com/CenterForOpenScience/osf.io/pull/5258#discussion_r57052684
 - [ ] use a list comprehension—https://github.com/CenterForOpenScience/osf.io/pull/5258#discussion_r57052793

File: api_tests/base/test_filters.py [hunk 3]
```diff
@@ -222,3 +229,51 @@ def test_convert_value_float(self):
...
+                return self.title
+        query_s = []
+        q1 = query('NewProj')
+        query_s.append((q1))
+        q2 = query('Zip')
+        query_s.append((q2))
+        q3 = query('Activity')
+        query_s.append((q3))
+        q4 = query('Activity')
+        query_s.append((q4))
```
Comment (whit537) [inline api_tests/base/test_filters.py]: You can greatly abbreviate this code by combining a couple Pythonic idioms:

```python
query_s = [query(x) for x in 'NewProj Zip Activity Activity'.split()]
```

File: api/base/filters.py [hunk 4]
```diff
@@ -24,17 +24,30 @@
...
+            if field[0] == '-':
+                field = field[1:]
+                a_field = getattr(a, field)
+                b_field = getattr(b, field)
+                if a_field > b_field:
+                    return -1
+                elif a_field < b_field:
+                    return 1
+                elif a_field == b_field:
+                    return 1
```
Comment (whit537) [inline api/base/filters.py]: Do you really mean `1` here? Usually when comparing, equality results in `0`, as in line 47.

Comment (whit537) [inline api/base/filters.py]: So, actually, the way this function had been implemented depends on the `0` case falling outside the loop. Why did you move the equality comparison inside the loop?

File: api_tests/base/test_filters.py [hunk 5]
```diff
@@ -222,3 +229,51 @@ def test_convert_value_float(self):
...
+        sort_q = sorted(query_s, cmp=filters.sort_multiple(['-title']))
+        sortt = []
+        for i in sort_q:
+            sortt.append(str(i))
+        assert_equal((sortt) , ['Zip', 'Proj', 'NewProj', 'Activity'])
+    
+    def test_filter_queryset_duplicate_fails(self):
+        class query:
```
Comment (whit537) [inline api_tests/base/test_filters.py]: Since we duplicate this helper class definition, let's move it up into the scope of the containing class, so we can access it as `self.query` in both tests.

File: api/base/filters.py (hunk 4 above)
Comment (whit537) [inline api/base/filters.py]: The (an?) error here is in sorting based on multiple fields; here's a field for a failing test (also illustrating the suggestion at https://github.com/CenterForOpenScience/osf.io/pull/5258#discussion_r57056259):

```
diff --git a/api_tests/base/test_filters.py b/api_tests/base/test_filters.py
index 7bcfefb..a9a4d4d 100644
--- a/api_tests/base/test_filters.py
+++ b/api_tests/base/test_filters.py
@@ -277,3 +277,15 @@ class TestODMOrderingFilter(ApiTestCase):
             sortt.append(str(i))
         assert_equal((sortt) , ['Zip', 'NewProj', 'Activity', 'Activity'])
 
+
+    class obj:
+        def __init__(self, **kw):
+            self.__dict__.update(**kw)
+
+    def test_filter_queryset_handles_multiple_fields(self):
+        objs = [self.obj(title='NewProj', n=10),
+                self.obj(title='Zip', n=20),
+                self.obj(title='Activity', n=30),
+                self.obj(title='Activity', n=40)]
+        actual = [x.n for x in sorted(objs, cmp=filters.sort_multiple(['title', '-number']))]
+        assert_equal(actual, [40, 30, 10, 20])
```

File: api/base/filters.py [hunk 6]
```diff
@@ -24,17 +24,30 @@
...
+            else:
+                a_field = getattr(a, field)
+                b_field = getattr(b, field)
+                if a_field > b_field:
+                    return 1
+                elif a_field < b_field:
+                    return -1
+                elif a_field == b_field:
+                    return 1
```
Comment (whit537) [inline api/base/filters.py]: I'd like to see [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)er code here. We should be able to implement this without duplicating the `getattr` calls, nor the `>` and `<` comparisons (hint: set a flag to `1` or `-1` based on the `field[0]` comparison and multiply the `return` value by that).

**Comment by whit537** [comment]
Ok @KAsante95, pass done and I've left you a [punchlist](https://github.com/CenterForOpenScience/osf.io/pull/5258#issuecomment-199978625) for you. Ping me when you're ready for more review! :-)

File: api_tests/base/test_filters.py [hunk 7]
```diff
@@ -222,3 +229,26 @@ def test_convert_value_float(self):
...
+        def __init__(self, title):
+            self.title = title
+        def __str__(self):
+            return self.title
+
+    def test_filter_queryset_fails(self):
+        query_s = [self.query(x) for x in 'NewProj Zip Proj Activity'.split()]
+        sort_q = sorted(query_s, cmp=filters.sort_multiple(['-title']))
+        sortt = [str(i) for i in sort_q]
+        assert_equal((sortt) , ['Zip', 'Proj', 'NewProj', 'Activity'])
```
Comment (whit537) [inline api_tests/base/test_filters.py]: We've still got extra parentheses around `sortt` here.

File: api/base/filters.py [hunk 8]
```diff
@@ -24,26 +24,19 @@
...
-                elif a_field == b_field:
-                    return 1
+            a_field = getattr(a, field)
+            b_field = getattr(b, field)
+            if a_field > b_field:
+                return 1 * rev
+            elif a_field < b_field:
+                return -1 * rev
+            elif a_field == b_field:
+                return 0
```
Comment (whit537) [inline api/base/filters.py]: This will still fail a test like the one at https://github.com/CenterForOpenScience/osf.io/pull/5258#discussion_r57059485. Let's go ahead and add that test to the test suite ... and then make it pass! :-)

**Comment by whit537** [comment]
[Punchlist](https://github.com/CenterForOpenScience/osf.io/pull/5258#issuecomment-199978625) updated! Back over to you, @KAsante95. :-)

File: api/base/filters.py [hunk 9]
```diff
@@ -24,20 +24,29 @@
 def sort_multiple(fields):
     fields = list(fields)
     def sort_fn(a, b):
-        while fields:
-            field = fields.pop(0)
+        rev = 1
+        i = 0
```
Comment (brianjgeiger) [inline api/base/filters.py]: You're not using `i` at all, so you should be able to remove it.

File: api/base/filters.py [hunk 10]
```diff
@@ -24,20 +24,29 @@
 def sort_multiple(fields):
     fields = list(fields)
     def sort_fn(a, b):
-        while fields:
-            field = fields.pop(0)
+        rev = 1
```
Comment (brianjgeiger) [inline api/base/filters.py]: Let's make rev something more readable. How about rename it into `sort_direction`, then define `FORWARD = 1` and `REVERSE = -1`. Then you can set `sort_direction = REVERSE` and `return 1 * sort_direction`, for example.

File: api_tests/base/test_filters.py [hunk 11]
```diff
@@ -222,3 +229,43 @@ def test_convert_value_float(self):
...
+        title = ' '
+        number = 0
+        def __init__(self, title, number):
+            self.title = title
+            self.number = number
+        def __str__(self):
+            return self.title
+
+    def test_filter_queryset_fails(self):
+        query_s = [self.query(x) for x in 'NewProj Zip Proj Activity'.split()]
```
Comment (brianjgeiger) [inline api_tests/base/test_filters.py]: I am not a big fan of vague variable names. Could you change `query_s`, `sort_q`, and `sortt` into things that are very clear as to their purpose? Same with anything else similar.

**Comment by brianjgeiger** [comment]
Make sure you have explicit tests for forward and reverse sorting, and also add an API view test to check explicitly for the 500 error from [OSF-5237]. You may have to try the test on the develop branch to see that it 500s there, then ensure that the test passes on your branch.

**Comment by brianjgeiger** [comment]
:penguin: Pass done.

**Comment by hmoco** [comment]
@sloria, as pointed out by Dawn, the bug I fixed in my PR seems to be fixed better and more throughly tested here. Could we see that this gets merged?

**Comment by sloria** [comment]
@hmoco Sure. Will merge this after it's through CR.

File: api_tests/base/test_filters.py [hunk 12]
```diff
@@ -222,3 +229,58 @@ def test_convert_value_float(self):
...
+            self.number = number
+        def __str__(self):
+            return self.title
+
+
+    def test_filter_queryset_forward(self):
+        query_to_be_sorted = [self.query(x) for x in 'NewProj Zip Proj Activity'.split()]
+        sorted_query = sorted(query_to_be_sorted, cmp=filters.sort_multiple(['title']))
+        sorted_output = [str(i) for i in sorted_query]
+        assert_equal((sorted_output) , ['Activity', 'NewProj', 'Proj', 'Zip'])
```
Comment (whit537) [inline api_tests/base/test_filters.py]: Why are we still using extra parens here?

File: api/base/filters.py [hunk 13]
```diff
@@ -24,20 +24,27 @@
...
-                return 1
+                return 1 * sort_direction
             elif a_field < b_field:
-                return -1
+                return -1 * sort_direction
+            elif a_field == b_field:
+                if index < len(fields) - 1:
+                    continue
+                else:
+                    return 0
```
Comment (whit537) [inline api/base/filters.py]: I believe the last conditional here is unnecessary (our tests pass without it); let's remove it.

That is, in the case where `index == len(fields) - 1`, we will be at the end of the `for` loop, and we'll hit the immediately subsequent `return 0`. In the `<` case, the explicit `continue` doesn't have any effect because we're already at the end of the `for` block; there are no additional statements that we're skipping by explicitly `continue`ing.

Comment (whit537) [inline api/base/filters.py]: (See also: https://github.com/CenterForOpenScience/osf.io/pull/5258#discussion_r57055035.)

**Comment by whit537** [comment]
[Punchlist](https://github.com/CenterForOpenScience/osf.io/pull/5258#issuecomment-199978625) updated! Back to you, @KAsante95 ... almost there! :-)

File: api/base/filters.py [hunk 14]
```diff
@@ -24,20 +24,24 @@
...
             a_field = getattr(a, field)
             b_field = getattr(b, field)
             if a_field > b_field:
-                return 1
+                return 1 * sort_direction
             elif a_field < b_field:
-                return -1
+                return -1 * sort_direction
+            elif a_field == b_field:
+                continue
```
Comment (whit537) [inline api/base/filters.py]: The last branch here is extraneous. Let's remove it.

File: api_tests/base/test_filters.py [hunk 15]
```diff
@@ -222,3 +229,58 @@ def test_convert_value_float(self):
...
+            self.number = number
+        def __str__(self):
+            return self.title
+
+
+    def test_filter_queryset_forward(self):
+        query_to_be_sorted = [self.query(x) for x in 'NewProj Zip Proj Activity'.split()]
+        sorted_query = sorted(query_to_be_sorted, cmp=filters.sort_multiple(['title']))
+        sorted_output = [str(i) for i in sorted_query]
+        assert_equal(sorted_output , ['Activity', 'NewProj', 'Proj', 'Zip'])
```
Comment (whit537) [inline api_tests/base/test_filters.py]: There shouldn't be a space between "`sorted_output`" and the comma.

Comment (whit537) [inline api_tests/base/test_filters.py]: (Sorry for not flagging this earlier.)

**Comment by whit537** [comment]
[Punchlist](https://github.com/CenterForOpenScience/osf.io/pull/5258#issuecomment-199978625) updated.

File: api/base/filters.py [hunk 16]
```diff
@@ -24,20 +24,23 @@
...
+                field = field[1:]
             a_field = getattr(a, field)
             b_field = getattr(b, field)
             if a_field > b_field:
-                return 1
+                return 1 * sort_direction
             elif a_field < b_field:
-                return -1
+                return -1 * sort_direction
+            continue
```
Comment (whit537) [inline api/base/filters.py]: `continue` is unnecessary at the end of a loop. It would only be used if you wanted to skip some additional code in the block of the loop. When there's nothing to skip, it's unneeded. Let's remove it! :-)

File: api/base/filters.py [hunk 17]
```diff
@@ -24,20 +24,22 @@
 def sort_multiple(fields):
     fields = list(fields)
     def sort_fn(a, b):
-        while fields:
-            field = fields.pop(0)
+        sort_direction = 1
+        for index, field in enumerate(fields):
```
Comment (whit537) [inline api/base/filters.py]: Gosh, I hate to keep nitpicking ... but we no longer need the enumeration here, since we're no longer using `index` as of be85a83ccbda1f2ee12ea7d42facd0d160a19cab.

I promise we're close, @KAsante95! ☺️

**Comment by whit537** [comment]
Okay! We're a little weak on the unit tests, but some is better than none. I'm good to go here. 👍
